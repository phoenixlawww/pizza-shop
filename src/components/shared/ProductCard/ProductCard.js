import PropTypes from 'prop-types'
import classnames from 'classnames'

import formatCurrency from 'utils/number/formatCurrency'

import Button, { BUTTON_VARIENTS } from 'components/shared/Button'

import './ProductCard.scss'

function ProductCard (props) {
  const {
    className,
    name,
    image,
    unitPrice,
    actionBtnLabel,
    onActionBtnClick
  } = props

  return (
    <div className={classnames('s-product-card', className)}>
      <div className="product-img-wrapper">
        {image && <img src={`${process.env.PUBLIC_URL}/${image}`} alt="" />}
      </div>
      <h4 className="product-name">{`${name} - $${formatCurrency(
        unitPrice
      )}`}</h4>
      <Button
        className="action-btn"
        label={actionBtnLabel}
        varient={BUTTON_VARIENTS.PRIMARY}
        onClick={(e) => {
          onActionBtnClick && onActionBtnClick()
        }}
      />
    </div>
  )
}

ProductCard.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  image: PropTypes.string,
  unitPrice: PropTypes.number.isRequired,
  actionBtnLabel: PropTypes.string,
  onActionBtnClick: PropTypes.func
}

ProductCard.defaultProps = {}

export default ProductCard
