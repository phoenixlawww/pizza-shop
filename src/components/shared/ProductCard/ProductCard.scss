@use "sass:math";
@import "~styles/tools";

.s-product-card {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: flex-start;
  padding: 10px;
  background-color: $white;
  border: 1px solid $light-grey;
  border-radius: 10px;

  > .product-img-wrapper {
    position: relative;
    overflow: hidden;
    background-color: $extra-light-grey;
    border-radius: 10px;

    &::before {
      display: block;
      width: 100%;
      padding-top: #{math.div(2, 3) * 100%};
      content: "";
    }

    > img {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      object-fit: cover;
    }
  }

  > .product-name {
    margin-bottom: 15px;
    text-align: center;
  }

  > .action-btn {
  }
}
