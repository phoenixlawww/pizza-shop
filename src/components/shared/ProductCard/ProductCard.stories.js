import ProductCard from './ProductCard'

import pizzaList from 'mock-data/products/pizza.json'

export default {
  component: ProductCard,
  title: 'Shared/ProductCard',
  args: {
    ...pizzaList.products[0],
    actionBtnLabel: 'Choose',
    onActionBtnClick: () => alert('clicked')
  },
  decorators: [
    (Story) => {
      return (
        <div style={{ maxWidth: 300 }}>
          <Story />
        </div>
      )
    }
  ]
}

const Template = (args) => <ProductCard {...args} />

export const Normal = Template.bind({})
Normal.args = {}

export const NoImage = Template.bind({})
NoImage.args = {
  image: ''
}
