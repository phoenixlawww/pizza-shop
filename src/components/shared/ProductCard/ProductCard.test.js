import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import ProductCard from './ProductCard'

import pizzaList from 'mock-data/products/pizza.json'

const actionBtnLabel = 'Choose'
const onActionBtnClick = jest.fn()
const testProps = {
  ...pizzaList.products[0],
  actionBtnLabel,
  onActionBtnClick
}

test('renders correctly', () => {
  const { baseElement } = render(<ProductCard {...testProps} />)

  expect(baseElement).toMatchSnapshot()
})

test('handle click event correctly', () => {
  render(<ProductCard {...testProps} />)

  const button = screen.getByText(actionBtnLabel)
  userEvent.click(button)

  expect(onActionBtnClick).toBeCalled()
})
