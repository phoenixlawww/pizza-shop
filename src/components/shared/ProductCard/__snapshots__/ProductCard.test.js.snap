// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<body>
  <div>
    <div
      class="s-product-card"
    >
      <div
        class="product-img-wrapper"
      >
        <img
          alt=""
          src="/pizza.jpg"
        />
      </div>
      <h4
        class="product-name"
      >
        Pizza A - $99
      </h4>
      <button
        class="s-button action-btn s-button-primary"
      >
        Choose
      </button>
    </div>
  </div>
</body>
`;
